test/common/testlib.py

```diff
@@ -739,6 +742,28 @@ def checkSuccess(self):
                 return False
         return True
 
+    def armTestTimeout(self):
+        assert self.test_timeout_orig_sighand is None
+
+        # most tests should take much less than 10mins, so default to that;
+        # longer tests can be annotated with @timeout(seconds)
+        # check the test function first, fall back to the class'es timeout
+        test_method = getattr(self.__class__, self._testMethodName)
+        timeout = getattr(test_method, "__timeout", getattr(self, "__timeout", 600))
+
+        def timeout_handler(signum, frame):
+            raise Error("%s timed out after %is" % (self, timeout))
+
+        self.test_timeout_orig_sighand = signal.signal(signal.SIGALRM, timeout_handler)
+        signal.alarm(timeout)
+
+    def disarmTestTimeout(self):
+        # this may be called multiple times
+        if self.test_timeout_orig_sighand is not None:
+            signal.alarm(0)
+            signal.signal(signal.SIGALRM, self.test_timeout_orig_sighand)
```

Comment:
What is the point of this line? `self.test_timeout_orig_sighand` is the previous signal handler, which always will be `timeout_handler`. Also you set `alarm(0)`, so you disabled the signals from coming. Maybe it would make more sense to set it up to `signal.SIG_IGN` if for some reason it need to be set up.

Reply:
No, `orig_sighand` is the initial value when the program started -- `signal.signal()` returns the *previous* handler when setting a new one, so that you can restore it properly. Usually that's `SIGDFL`, but that's fungible -- I'd rather not make assumptions there, and just restore it to  the previous value.

Indeed with `alarm(0)` we don't strictly have to reset it back, but in principle other parts of the code may use (short-lived) alarms as well.

Reply:
ah right, makes sense! Thanks for the explanation :)